### PR TITLE
PWX-30576: Create clusterpair to support secret and backuplocations creation with required cloud credentials.

### PR DESCRIPTION
**Pull request by diptiranjanpx**
Signed-Off-By: Diptiranjan

**What type of PR is this?**
>improvement

**What this PR does / why we need it**:
Clusterpair creation to create credential secret and backuplocation for use.

**Does this PR change a user-facing CRD or CLI?**:
yes
```
Example with s3 objectstore.

➜  ~ ./storkctl create clusterpair cp-new -p s3 -n kube-system --s3-access-key ***** --s3-secret-key ***** --s3-region us-east-1 --s3-endpoint *******
 --src-kube-file /tmp/src.config --dest-kube-file /tmp/dest.config --src-ip ****** --dest-ip ********
Using PX-Service Endpoint of DR cluster to create clusterpair...

Creating PX-Service Endpoint of DR cluster to create clusterpair...

Backuplocation cp-new added successfully on source cluster in namespace kube-system
ClusterPair cp-new created successfully on source cluster
Using PX-Service endpoints of source cluster to create clusterpair...

Backuplocation cp-new added successfully on destination cluster in namespace kube-system
ClusterPair cp-new created successfully on destination cluster

In source :
➜  ~ ./storkctl get clusterpair -n kube-system
NAME          STORAGE-STATUS   SCHEDULER-STATUS   CREATED
cp-new        Ready            Ready              31 May 23 12:31 UTC

In dest:
➜  ~  ./storkctl get clusterpair -n kube-system
NAME          STORAGE-STATUS   SCHEDULER-STATUS   CREATED
cp-new        Ready            Ready              31 May 23 12:31 UTC

MigrationSchedule using the clusterpair “cp-new”
➜  ~ ./storkctl get migrations -n kube-system
NAME                                             CLUSTERPAIR   STAGE   STATUS       VOLUMES   RESOURCES   CREATED               ELAPSED                           TOTAL BYTES TRANSFERRED
pwx-29789-mongo-interval-2023-05-31-131743       cp-new        Final   Successful   6/6       32/32       31 May 23 13:17 UTC   Volumes (1m37s) Resources (22s)   1022590976
```

**Is a release note needed?**:

**Does this change need to be cherry-picked to a release branch?**:

### CODE REVIEW

File: pkg/storkctl/clusterpair.go [hunk 1]
```diff
@@ -258,6 +266,84 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+				credentialData["accountKey"] = keyData
+				credentialData["projectID"] = []byte(googleProjectID)
+				backupLocation.Location.Type = storkv1.BackupLocationGoogle
+			default:
+				util.CheckErr(fmt.Errorf("provider needs to be either azure, google, s3"))
+				return
+			}
+
+			credentialData["path"] = []byte(bucket)
+			credentialData["type"] = []byte(provider)
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:344]: Add a check if provider is one of s3, azure or google

File: pkg/storkctl/clusterpair.go [hunk 2]
```diff
@@ -341,6 +474,24 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
 	createClusterPairCommand.Flags().StringVarP(&destToken, "dest-token", "", "", "(Optional)Destination cluster token for cluster pairing")
 	createClusterPairCommand.Flags().StringVarP(&projectMappingsStr, "project-mappings", "", "", projectMappingHelpString)
 
+	// New parameters for creating backuplocation secret
+	createClusterPairCommand.Flags().StringVarP(&provider, "provider", "p", "", "Cloud provider type [s3,azure,google]")
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:478]:
```suggestion
	createClusterPairCommand.Flags().StringVarP(&provider, "provider", "p", "", "External objectstore provider name. [s3, azure, google]")
```

File: pkg/storkctl/clusterpair.go [hunk 3]
```diff
@@ -341,6 +474,24 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+	createClusterPairCommand.Flags().StringVar(&bucket, "bucket", "", "Bucket name")
+	createClusterPairCommand.Flags().StringVar(&encryptionKey, "encryption-key", "", "Encryption key for backuplocation")
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:480]:
```suggestion
	createClusterPairCommand.Flags().StringVar(&encryptionKey, "encryption-key", "", "Encryption key for encrypting the data stored in the objectstore.")
```

File: pkg/storkctl/clusterpair.go [hunk 4]
```diff
@@ -341,6 +474,24 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+	// AWS
+	createClusterPairCommand.Flags().StringVar(&s3AccessKey, "s3-access-key", "", "Accesskey for AWS")
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:482]:
```suggestion
	createClusterPairCommand.Flags().StringVar(&s3AccessKey, "s3-access-key", "", "AWS Access Key [AWS_ACCESS_KEY]")
```

File: pkg/storkctl/clusterpair.go [hunk 5]
```diff
@@ -341,6 +474,24 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+	createClusterPairCommand.Flags().StringVar(&s3SecretKey, "s3-secret-key", "", "Secret key for AWS")
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:483]:
```suggestion
	createClusterPairCommand.Flags().StringVar(&s3SecretKey, "s3-secret-key", "", "AWS Secret key [AWS_SECRET_KEY]")
```

File: pkg/storkctl/clusterpair.go [hunk 6]
```diff
@@ -341,6 +474,24 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+	createClusterPairCommand.Flags().StringVar(&s3Region, "s3-region", "", "Region for AWS")
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:484]: Can you follow the same suite for all the aws parameters? Also specify the env variable standard spec

File: pkg/storkctl/clusterpair.go [hunk 7]
```diff
@@ -341,6 +474,24 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+	createClusterPairCommand.Flags().StringVar(&s3EndPoint, "s3-endpoint", "", "Endpoint for AWS")
+	createClusterPairCommand.Flags().StringVar(&s3StorageClass, "s3-storage-class", "", "Storage Class for AWS")
+	createClusterPairCommand.Flags().BoolVar(&disableSSL, "disable-ssl", false, "Disable for ssl for s3. ex: --disable-ssl=true, enable: --disable-ssl=false")
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:487]:
```suggestion
	createClusterPairCommand.Flags().BoolVar(&disableSSL, "disable-ssl", false, "Set to true to disable ssl when using s3")
```

File: pkg/storkctl/clusterpair.go [hunk 8]
```diff
@@ -285,7 +371,39 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
 				util.CheckErr(err)
 				return
 			}
+
+			printMsg("Creating PX-Service Endpoint of DR cluster to create clusterpair...\n", ioStreams.Out)
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:375]: log should be Creating Secrets and BackupLocations ?

**Review by adityadani — COMMENTED** [review]
overall looks good. Can you change at least 1 migration integration test to use `storkctl` to create a cluster pair using the new commands?

File: pkg/storkctl/clusterpair.go (hunk 1 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go:344]: The check was there , modified the message though.

File: pkg/storkctl/clusterpair.go (hunk 2 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go (hunk 3 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go (hunk 4 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: As discussed modified it to "Access Key for S3"

File: pkg/storkctl/clusterpair.go (hunk 5 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: As discussed modified it to "Secret Key for S3"

File: pkg/storkctl/clusterpair.go (hunk 6 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: As discussed, not putting the env vars for each parameter.

File: pkg/storkctl/clusterpair.go (hunk 7 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go (hunk 8 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go [hunk 9]
```diff
@@ -27,10 +28,11 @@ import (
...
 const (
-	clusterPairSubcommand = "clusterpair"
-	cmdPathKey            = "cmd-path"
-	gcloudPath            = "./google-cloud-sdk/bin/gcloud"
-	gcloudBinaryName      = "gcloud"
+	clusterPairSubcommand  = "clusterpair"
+	cmdPathKey             = "cmd-path"
+	gcloudPath             = "./google-cloud-sdk/bin/gcloud"
+	gcloudBinaryName       = "gcloud"
+	skipResourceAnnotation = "stork.libopenstorage.org/skip-resource"
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:35]: nit: Can you define this as a global constant somewhere and change all the references in cli and controllers?

File: pkg/storkctl/clusterpair.go [hunk 10]
```diff
@@ -258,7 +266,84 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+				keyData, err := os.ReadFile(googleJSONKey)
+				if err != nil {
+					util.CheckErr(fmt.Errorf("failed to read json key file: %v", err))
+					return
+				}
+				credentialData["accountKey"] = keyData
+				credentialData["projectID"] = []byte(googleProjectID)
+				backupLocation.Location.Type = storkv1.BackupLocationGoogle
+			default:
+				util.CheckErr(getMissingParameterError("provider", "External objectstore Provider needs to be either of azure, google, s3"))
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:339]: nit: lower case provider

File: pkg/storkctl/clusterpair.go [hunk 11]
```diff
@@ -285,17 +370,50 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+			annotations[skipResourceAnnotation] = "true"
+			_, err = storkops.Instance().CreateBackupLocation(backupLocation)
+			if err != nil {
+				err := fmt.Errorf("unable to create backuplocation in source. Err: %v", err)
+				util.CheckErr(err)
+				return
+			}
+			printMsg(fmt.Sprintf("Backuplocation %v added successfully on source cluster in namespace %v\n", clusterPairName, cmdFactory.GetNamespace()), ioStreams.Out)
+
+			printMsg("Using PX-Service Endpoint of destination cluster to create clusterpair on source cluster...", ioStreams.Out)
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:405]: nit: lets change this to : Creating a cluster pair. Direction: Source -> Destination

File: pkg/storkctl/clusterpair.go [hunk 12]
```diff
@@ -285,17 +370,50 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+			srcClusterPair.Spec.Options[storkv1.BackupLocationResourceName] = backupLocation.Name
 			_, err = storkops.Instance().CreateClusterPair(srcClusterPair)
 			if err != nil {
 				util.CheckErr(err)
 				return
 			}
-			printMsg("ClusterPair "+clusterPairName+" created successfully on source cluster", ioStreams.Out)
+			printMsg(fmt.Sprintf("ClusterPair %s created successfully on source cluster in namespace %v\n", clusterPairName, cmdFactory.GetNamespace()), ioStreams.Out)
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:412]: nit: Cluster pair created successfully. Direction: Source -> Destination

File: pkg/storkctl/clusterpair.go [hunk 13]
```diff
@@ -321,13 +439,30 @@ func newCreateClusterPairCommand(cmdFactory Factory, ioStreams genericclioptions
...
+			_, err = storkops.Instance().CreateBackupLocation(backupLocation)
 			if err != nil {
+				err := fmt.Errorf("unable to create backuplocation in destination. Err: %v", err)
 				util.CheckErr(err)
 				return
 			}
-			printMsg("ClusterPair "+clusterPairName+" created successfully on destination cluster", ioStreams.Out)
+			printMsg(fmt.Sprintf("Backuplocation %v added successfully on destination cluster in namespace %v\n", clusterPairName, cmdFactory.GetNamespace()), ioStreams.Out)
 
+			printMsg("Using PX-Service endpoints of source cluster to create clusterpair on destination cluster...", ioStreams.Out)
```
Comment (adityadani) [inline pkg/storkctl/clusterpair.go:458]: in a similar fashion change the messages here

**Review by adityadani — APPROVED** [review]
lgtm

File: pkg/storkctl/clusterpair.go (hunk 11 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go (hunk 12 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go (hunk 13 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go (hunk 10 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go]: Done

File: pkg/storkctl/clusterpair.go (hunk 9 above)
Comment (diptiranjanpx) [reply pkg/storkctl/clusterpair.go:35]: Not implementing this for now as it this constant is also used outside of packages in volumes also. Will implement in another PR.

**Pull request merged**